In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
import kagglehub
path=kagglehub.dataset_download('vijayuv/onlineretail')

Using Colab cache for faster access to the 'onlineretail' dataset.


In [ ]:
onre=pd.read_csv('/kaggle/input/onlineretail/OnlineRetail.csv',encoding='ISO-8859-1')

#Task 1: Load the dataset and display basic information about its structure

In [ ]:
onre.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


Task 2: Show the first 15 and last 15 records

In [ ]:
onre.head(15)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,12/1/2010 8:26,7.65,17850.0,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,12/1/2010 8:26,4.25,17850.0,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,12/1/2010 8:28,1.85,17850.0,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,12/1/2010 8:28,1.85,17850.0,United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,12/1/2010 8:34,1.69,13047.0,United Kingdom


In [ ]:
onre.tail(15)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541894,581587,22631,CIRCUS PARADE LUNCH BOX,12,12/9/2011 12:50,1.95,12680.0,France
541895,581587,22556,PLASTERS IN TIN CIRCUS PARADE,12,12/9/2011 12:50,1.65,12680.0,France
541896,581587,22555,PLASTERS IN TIN STRONGMAN,12,12/9/2011 12:50,1.65,12680.0,France
541897,581587,22728,ALARM CLOCK BAKELIKE PINK,4,12/9/2011 12:50,3.75,12680.0,France
541898,581587,22727,ALARM CLOCK BAKELIKE RED,4,12/9/2011 12:50,3.75,12680.0,France
541899,581587,22726,ALARM CLOCK BAKELIKE GREEN,4,12/9/2011 12:50,3.75,12680.0,France
541900,581587,22730,ALARM CLOCK BAKELIKE IVORY,4,12/9/2011 12:50,3.75,12680.0,France
541901,581587,22367,CHILDRENS APRON SPACEBOY DESIGN,8,12/9/2011 12:50,1.95,12680.0,France
541902,581587,22629,SPACEBOY LUNCH BOX,12,12/9/2011 12:50,1.95,12680.0,France
541903,581587,23256,CHILDRENS CUTLERY SPACEBOY,4,12/9/2011 12:50,4.15,12680.0,France


Task 3: Display a random sample of 20 transactions.

In [ ]:
onre.sample(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
140008,548373,21731,RED TOADSTOOL LED NIGHT LIGHT,2,3/30/2011 16:34,3.29,NaN,United Kingdom
442478,574673,23206,LUNCH BAG APPLE DESIGN,3,11/6/2011 12:26,1.65,15555.0,United Kingdom
495272,578295,37450,CERAMIC CAKE BOWL + HANGING CAKES,6,11/23/2011 14:39,2.95,12569.0,Germany
349270,567484,82484,WOOD BLACK BOARD ANT WHITE FINISH,12,9/20/2011 13:52,6.75,17340.0,United Kingdom
396332,571061,82580,BATHROOM METAL SIGN,1,10/13/2011 14:33,1.25,NaN,United Kingdom
92589,544201,35004B,SET OF 3 BLACK FLYING DUCKS,1,2/17/2011 10:11,12.46,NaN,United Kingdom
223339,556481,22674,FRENCH TOILET SIGN BLUE METAL,2,6/12/2011 13:00,1.25,16115.0,United Kingdom
168595,551085,22938,CUPCAKE LACE PAPER SET 6,12,4/26/2011 12:29,1.95,14854.0,United Kingdom
287227,C562097,82483,WOOD 2 DRAWER CABINET WHITE FINISH,-1,8/2/2011 13:19,5.95,15311.0,United Kingdom
361110,568327,23231,WRAP VINTAGE DOILY,25,9/26/2011 14:02,0.42,15492.0,United Kingdom


Task 4: Find out how many rows and columns are in the dataset.

In [ ]:
onre.shape

(541909, 8)

Task 5: List all column names.

In [ ]:
onre.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

Task 6: Check the data type of each column.

In [ ]:
onre.dtypes

,0
InvoiceNo,object
StockCode,object
Description,object
Quantity,int64
InvoiceDate,object
UnitPrice,float64
CustomerID,float64
Country,object


Task 7: Get a statistical summary of all numerical columns

In [ ]:
onre.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


#PART B: Data Quality Assessment


Task 8: Identify which columns have missing values and how many

In [ ]:
onre.isnull().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


Task 9: Find out how many transactions have missing CustomerID

In [ ]:
onre['CustomerID'].isnull().sum()

np.int64(135080)

Task 10: Check if there are any duplicate rows in the dataset.

In [ ]:
onre.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
541904,False
541905,False
541906,False
541907,False


In [ ]:
onre.duplicated().sum()

np.int64(5268)

Task 11: Identify transactions where Quantity is less than or equal to zero.

In [ ]:
onre[onre['Quantity']<=0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


Task 12: Find transactions where UnitPrice is zero or negative.

In [ ]:
onre[(onre['UnitPrice']==0) | (onre['UnitPrice']<0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,12/8/2011 13:58,0.0,NaN,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,12/8/2011 13:58,0.0,NaN,United Kingdom
538554,581408,85175,NaN,20,12/8/2011 14:06,0.0,NaN,United Kingdom


Task 13: Identify how many invoice numbers start with the letter 'C'.

In [ ]:
onre['InvoiceNo'].str.startswith('C').sum()

np.int64(9288)

Task 14: Check if the Description column has any leading or trailing spaces.

In [ ]:
onre.columns.str.startswith(' ') | onre.columns.str.endswith(' ')

array([False, False, False, False, False, False, False, False])

#PART C: Data Cleaning

Task 15: Remove all rows where CustomerID is missing

In [ ]:
onre['CustomerID'].dropna()


,CustomerID
0,17850.0
1,17850.0
2,17850.0
3,17850.0
4,17850.0
...,...
541904,12680.0
541905,12680.0
541906,12680.0
541907,12680.0


Task 16: For missing product descriptions, fill them with "Unknown Product"

In [ ]:
onre['Description']=onre['Description'].fillna('Unknown Product')
onre['Description']

,Description
0,WHITE HANGING HEART T-LIGHT HOLDER
1,WHITE METAL LANTERN
2,CREAM CUPID HEARTS COAT HANGER
3,KNITTED UNION FLAG HOT WATER BOTTLE
4,RED WOOLLY HOTTIE WHITE HEART.
...,...
541904,PACK OF 20 SPACEBOY NAPKINS
541905,CHILDREN'S APRON DOLLY GIRL
541906,CHILDRENS CUTLERY DOLLY GIRL
541907,CHILDRENS CUTLERY CIRCUS PARADE


Task 17: Remove all duplicate transactions.

In [ ]:
onre

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


In [ ]:
onre.drop_duplicates()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


Task 18: Remove transactions where Quantity is less than or equal to zero.

In [ ]:
onre=onre.drop(onre[onre['Quantity']<=0].index)
onre

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


Task 19: Remove transactions where UnitPrice is less than or equal to zero

In [ ]:
onre=onre.drop(onre[onre['UnitPrice']<=0].index)
onre

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


Task 20: Remove all cancelled transactions (invoices starting with 'C').

In [ ]:
onre.drop(onre[onre['InvoiceNo'].str.startswith('C')].index,inplace=True)

Task 21: Change the column name "Description" to "ProductName".

In [ ]:
onre.rename(columns={'Description': 'ProductName'},inplace=True)

Task 22: Ensure the InvoiceDate column is in proper date-time format.

In [ ]:
onre['InvoiceDate']=pd.to_datetime(onre['InvoiceDate'])

In [ ]:
onre['InvoiceDate'].dtype

dtype('<M8[ns]')

Task 23: Convert CustomerID to integer type.

In [ ]:
onre['CustomerID'].astype(float)

,CustomerID
0,17850.0
1,17850.0
2,17850.0
3,17850.0
4,17850.0
...,...
541904,12680.0
541905,12680.0
541906,12680.0
541907,12680.0


Task 24: Reset the index after all cleaning operations.

In [ ]:
onre.reset_index(drop=True,inplace=True)

#PART D: Feature Creation

Task 25: Create a new column called "TotalPrice" that represents the total value of each transaction

In [ ]:
onre['TotalPrice']=onre['Quantity']*onre['UnitPrice']
onre.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 530104 entries, 0 to 530103
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    530104 non-null  object        
 1   StockCode    530104 non-null  object        
 2   ProductName  530104 non-null  object        
 3   Quantity     530104 non-null  int64         
 4   InvoiceDate  530104 non-null  datetime64[ns]
 5   UnitPrice    530104 non-null  float64       
 6   CustomerID   397884 non-null  float64       
 7   Country      530104 non-null  object        
 8   TotalPrice   530104 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(4)
memory usage: 36.4+ MB


Task 26: Extract the year from InvoiceDate and store it in a new column "Year"

In [ ]:
onre['Year']=onre['InvoiceDate'].dt.year
onre['Year']

,Year
0,2010
1,2010
2,2010
3,2010
4,2010
...,...
530099,2011
530100,2011
530101,2011
530102,2011


Task 27: Extract the month from InvoiceDate and store it in a new column "Month".

In [ ]:
onre['Month']=onre['InvoiceDate'].dt.month
onre['Month']

,Month
0,12
1,12
2,12
3,12
4,12
...,...
530099,12
530100,12
530101,12
530102,12


Task 28: Extract the day from InvoiceDate and store it in a new column "Day".

In [ ]:
onre['Day']=onre['InvoiceDate'].dt.day
onre['Day']

,Day
0,1
1,1
2,1
3,1
4,1
...,...
530099,9
530100,9
530101,9
530102,9


Task 29: Extract the hour from InvoiceDate and store it in a new column "Hour".

In [ ]:
onre['Hour']=onre['InvoiceDate'].dt.hour
onre['Hour']

,Hour
0,8
1,8
2,8
3,8
4,8
...,...
530099,12
530100,12
530101,12
530102,12


Task 30: Create a column "DayOfWeek" showing which day of the week the transaction occurred.

In [ ]:
onre['DayOfWeek']=onre['InvoiceDate'].dt.day
onre['DayOfWeek']

,DayOfWeek
0,1
1,1
2,1
3,1
4,1
...,...
530099,9
530100,9
530101,9
530102,9


Task 31: Create a column "Quarter" showing which quarter (Q1, Q2, Q3, Q4) the transaction belongs to

In [ ]:
onre['Quarter']='Q' + onre['InvoiceDate'].dt.quarter.astype(str)
onre['Quarter']

,Quarter
0,Q4
1,Q4
2,Q4
3,Q4
4,Q4
...,...
530099,Q4
530100,Q4
530101,Q4
530102,Q4


Task 32: Create a column "RevenueCategory" with three levels:

"Low" for transactions under £10
"Medium" for transactions between £10 and £100
"High" for transactions above £100

In [ ]:
onre['RevenueCategory']=onre['TotalPrice'].apply(lambda x: 'Low' if x<10 else('Medium' if x<=100 else 'High'))
onre['RevenueCategory']

,RevenueCategory
0,Medium
1,Medium
2,Medium
3,Medium
4,Medium
...,...
530099,Medium
530100,Medium
530101,Medium
530102,Medium


Task 33: Create a column "OrderSize" with three levels:

"Small" for Quantity less than 5
"Medium" for Quantity between 5 and 20
"Large" for Quantity greater than 20

In [ ]:
onre['OrderSize'] = onre['Quantity'].apply(lambda x: 'Small' if x < 5 else ('Medium' if x <= 20 else 'Large'))
onre['OrderSize']

,OrderSize
0,Medium
1,Medium
2,Medium
3,Medium
4,Medium
...,...
530099,Medium
530100,Medium
530101,Small
530102,Small


Task 34: Create a column that shows the length of each product name.

In [ ]:
onre['LenOfProductName']=onre['ProductName'].str.len()
onre['LenOfProductName']

,LenOfProductName
0,34
1,19
2,30
3,35
4,30
...,...
530099,27
530100,28
530101,29
530102,31


PART E: Text Data Processing

Task 35: Convert all product names to uppercase.

In [ ]:
onre['ProductName'] = onre['ProductName'].str.upper()
onre['ProductName']

,ProductName
0,WHITE HANGING HEART T-LIGHT HOLDER
1,WHITE METAL LANTERN
2,CREAM CUPID HEARTS COAT HANGER
3,KNITTED UNION FLAG HOT WATER BOTTLE
4,RED WOOLLY HOTTIE WHITE HEART.
...,...
530099,PACK OF 20 SPACEBOY NAPKINS
530100,CHILDREN'S APRON DOLLY GIRL
530101,CHILDRENS CUTLERY DOLLY GIRL
530102,CHILDRENS CUTLERY CIRCUS PARADE


Task 36: Remove any extra spaces from the Country column.

In [ ]:
onre['Country']=onre['Country'].str.strip()
onre['Country']

,Country
0,United Kingdom
1,United Kingdom
2,United Kingdom
3,United Kingdom
4,United Kingdom
...,...
530099,France
530100,France
530101,France
530102,France


Task 37: Find all products that contain the word "BAG" in their name.

In [ ]:
onre[onre['ProductName'].str.upper().str.contains('BAG')]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
38,536370,22661,CHARLOTTE BAG DOLLY GIRL DESIGN,20,2010-12-01 08:45:00,0.85,12583.0,France,17.0,2010,12,1,8,1,Q4,Medium,Medium,31
86,536378,22386,JUMBO BAG PINK POLKADOT,10,2010-12-01 09:37:00,1.95,14688.0,United Kingdom,19.5,2010,12,1,9,1,Q4,Medium,Medium,23
87,536378,85099C,JUMBO BAG BAROQUE BLACK WHITE,10,2010-12-01 09:37:00,1.95,14688.0,United Kingdom,19.5,2010,12,1,9,1,Q4,Medium,Medium,30
88,536378,21033,JUMBO BAG CHARLIE AND LOLA TOYS,10,2010-12-01 09:37:00,2.95,14688.0,United Kingdom,29.5,2010,12,1,9,1,Q4,Medium,Medium,31
89,536378,20723,STRAWBERRY CHARLOTTE BAG,10,2010-12-01 09:37:00,0.85,14688.0,United Kingdom,8.5,2010,12,1,9,1,Q4,Low,Medium,24
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
530028,581579,23343,JUMBO BAG VINTAGE CHRISTMAS,30,2011-12-09 12:19:00,1.79,17581.0,United Kingdom,53.7,2011,12,9,12,9,Q4,Medium,Large,28
530029,581579,23344,JUMBO BAG 50'S CHRISTMAS,20,2011-12-09 12:19:00,1.79,17581.0,United Kingdom,35.8,2011,12,9,12,9,Q4,Medium,Medium,25
530030,581579,23581,JUMBO BAG PAISLEY PARK,40,2011-12-09 12:19:00,1.79,17581.0,United Kingdom,71.6,2011,12,9,12,9,Q4,Medium,Large,22
530057,581581,23681,LUNCH BAG RED VINTAGE DOILY,10,2011-12-09 12:20:00,1.65,17581.0,United Kingdom,16.5,2011,12,9,12,9,Q4,Medium,Medium,27


Task 38: Find all products whose names start with "WHITE".

In [ ]:
onre[onre['ProductName'].str.upper().str.startswith('WHITE')]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30,2010,12,1,8,1,Q4,Medium,Medium,34
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,1,8,1,Q4,Medium,Medium,19
49,536373,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 09:02:00,2.55,17850.0,United Kingdom,15.30,2010,12,1,9,1,Q4,Medium,Medium,34
50,536373,71053,WHITE METAL LANTERN,6,2010-12-01 09:02:00,3.39,17850.0,United Kingdom,20.34,2010,12,1,9,1,Q4,Medium,Medium,19
66,536375,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 09:32:00,2.55,17850.0,United Kingdom,15.30,2010,12,1,9,1,Q4,Medium,Medium,34
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529478,581497,21673,WHITE SPOT BLUE CERAMIC DRAWER KNOB,1,2011-12-09 10:23:00,3.29,NaN,United Kingdom,3.29,2011,12,9,10,9,Q4,Low,Small,35
529710,581498,71053,WHITE MOROCCAN METAL LANTERN,1,2011-12-09 10:26:00,8.29,NaN,United Kingdom,8.29,2011,12,9,10,9,Q4,Low,Small,28
529777,581502,23210,WHITE ROCKING HORSE HAND PAINTED,12,2011-12-09 10:51:00,1.25,15910.0,United Kingdom,15.00,2011,12,9,10,9,Q4,Medium,Medium,32
529845,581538,21673,WHITE SPOT BLUE CERAMIC DRAWER KNOB,1,2011-12-09 11:34:00,1.45,14446.0,United Kingdom,1.45,2011,12,9,11,9,Q4,Low,Small,35


Task 39: Find all products whose names end with "HOLDER".

In [ ]:
onre[onre['ProductName'].str.upper().str.endswith('HOLDER')]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.3,2010,12,1,8,1,Q4,Medium,Medium,34
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,17850.0,United Kingdom,25.5,2010,12,1,8,1,Q4,Medium,Medium,33
49,536373,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 09:02:00,2.55,17850.0,United Kingdom,15.3,2010,12,1,9,1,Q4,Medium,Medium,34
64,536373,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 09:02:00,4.25,17850.0,United Kingdom,25.5,2010,12,1,9,1,Q4,Medium,Medium,33
66,536375,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 09:32:00,2.55,17850.0,United Kingdom,15.3,2010,12,1,9,1,Q4,Medium,Medium,34
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529909,581567,21326,AGED GLASS SILVER T-LIGHT HOLDER,144,2011-12-09 11:56:00,0.55,16626.0,United Kingdom,79.2,2011,12,9,11,9,Q4,Medium,Large,32
529936,581571,85053,FRENCH ENAMEL CANDLEHOLDER,2,2011-12-09 12:00:00,2.10,15311.0,United Kingdom,4.2,2011,12,9,12,9,Q4,Low,Small,26
530020,581579,22197,POPCORN HOLDER,24,2011-12-09 12:19:00,0.85,17581.0,United Kingdom,20.4,2011,12,9,12,9,Q4,Medium,Large,14
530067,581585,22460,EMBOSSED GLASS TEALIGHT HOLDER,12,2011-12-09 12:31:00,1.25,15804.0,United Kingdom,15.0,2011,12,9,12,9,Q4,Medium,Medium,30


Task 40: Replace "United Kingdom" with "UK" in the Country column.

In [ ]:
onre['Country']=onre['Country'].str.replace('United Kingdom','UK')

Task 41: Replace "EIRE" with "Ireland" in the Country column.

In [ ]:
onre['Country']=onre['Country'].str.replace('Ireland','EIRE')

#PART F: Data Selection and Filtering

Task 42: Select only the columns: InvoiceNo, ProductName, Quantity, UnitPrice, Country

In [ ]:
onre[['InvoiceNo','ProductName','Quantity','UnitPrice','Country']]

,InvoiceNo,ProductName,Quantity,UnitPrice,Country
0,536365,WHITE HANGING HEART T-LIGHT HOLDER,6,2.55,UK
1,536365,WHITE METAL LANTERN,6,3.39,UK
2,536365,CREAM CUPID HEARTS COAT HANGER,8,2.75,UK
3,536365,KNITTED UNION FLAG HOT WATER BOTTLE,6,3.39,UK
4,536365,RED WOOLLY HOTTIE WHITE HEART.,6,3.39,UK
...,...,...,...,...,...
530099,581587,PACK OF 20 SPACEBOY NAPKINS,12,0.85,France
530100,581587,CHILDREN'S APRON DOLLY GIRL,6,2.10,France
530101,581587,CHILDRENS CUTLERY DOLLY GIRL,4,4.15,France
530102,581587,CHILDRENS CUTLERY CIRCUS PARADE,4,4.15,France


Task 43: Display transactions from Germany only.

In [ ]:
onre[onre['Country']=='Germany']

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
1098,536527,22809,SET OF 6 T-LIGHTS SANTA,6,2010-12-01 13:04:00,2.95,12662.0,Germany,17.70,2010,12,1,13,1,Q4,Medium,Medium,23
1099,536527,84347,ROTATING SILVER ANGELS T-LIGHT HLDR,6,2010-12-01 13:04:00,2.55,12662.0,Germany,15.30,2010,12,1,13,1,Q4,Medium,Medium,35
1100,536527,84945,MULTI COLOUR SILVER T-LIGHT HOLDER,12,2010-12-01 13:04:00,0.85,12662.0,Germany,10.20,2010,12,1,13,1,Q4,Medium,Medium,34
1101,536527,22242,5 HOOK HANGER MAGIC TOADSTOOL,12,2010-12-01 13:04:00,1.65,12662.0,Germany,19.80,2010,12,1,13,1,Q4,Medium,Medium,29
1102,536527,22244,3 HOOK HANGER MAGIC GARDEN,12,2010-12-01 13:04:00,1.95,12662.0,Germany,23.40,2010,12,1,13,1,Q4,Medium,Medium,26
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529996,581578,22993,SET OF 4 PANTRY JELLY MOULDS,12,2011-12-09 12:16:00,1.25,12713.0,Germany,15.00,2011,12,9,12,9,Q4,Medium,Medium,28
529997,581578,22907,PACK OF 20 NAPKINS PANTRY DESIGN,12,2011-12-09 12:16:00,0.85,12713.0,Germany,10.20,2011,12,9,12,9,Q4,Medium,Medium,32
529998,581578,22908,PACK OF 20 NAPKINS RED APPLES,12,2011-12-09 12:16:00,0.85,12713.0,Germany,10.20,2011,12,9,12,9,Q4,Medium,Medium,29
529999,581578,23215,JINGLE BELL HEART ANTIQUE SILVER,12,2011-12-09 12:16:00,2.08,12713.0,Germany,24.96,2011,12,9,12,9,Q4,Medium,Medium,32


Task 44: Display transactions where Quantity is greater than 50

In [ ]:
onre[onre['Quantity']>=50]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
46,536371,22086,PAPER CHAIN KIT 50'S CHRISTMAS,80,2010-12-01 09:00:00,2.55,13748.0,UK,204.00,2010,12,1,9,1,Q4,High,Large,31
83,536376,21733,RED HANGING HEART T-LIGHT HOLDER,64,2010-12-01 09:32:00,2.55,15291.0,UK,163.20,2010,12,1,9,1,Q4,High,Large,32
96,536378,21212,PACK OF 72 RETROSPOT CAKE CASES,120,2010-12-01 09:37:00,0.42,14688.0,UK,50.40,2010,12,1,9,1,Q4,Medium,Large,31
102,536378,85071B,RED CHARLIE+LOLA PERSONAL DOORSIGN,96,2010-12-01 09:37:00,0.38,14688.0,UK,36.48,2010,12,1,9,1,Q4,Medium,Large,34
146,536382,22381,TOY TIDY PINK POLKADOT,50,2010-12-01 09:45:00,1.85,16098.0,UK,92.50,2010,12,1,9,1,Q4,Medium,Large,22
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529900,581566,23404,HOME SWEET HOME BLACKBOARD,144,2011-12-09 11:50:00,3.26,18102.0,UK,469.44,2011,12,9,11,9,Q4,High,Large,26
529901,581567,21417,COCKLE SHELL DISH,84,2011-12-09 11:56:00,0.79,16626.0,UK,66.36,2011,12,9,11,9,Q4,Medium,Large,17
529909,581567,21326,AGED GLASS SILVER T-LIGHT HOLDER,144,2011-12-09 11:56:00,0.55,16626.0,UK,79.20,2011,12,9,11,9,Q4,Medium,Large,32
529941,581571,23167,SMALL CERAMIC TOP STORAGE JAR,96,2011-12-09 12:00:00,0.69,15311.0,UK,66.24,2011,12,9,12,9,Q4,Medium,Large,30


Task 45: Display transactions where UnitPrice is between £5 and £10.

In [ ]:
onre[(onre['UnitPrice']>=5) & (onre['UnitPrice']<=10)]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,17850.0,UK,15.30,2010,12,1,8,1,Q4,Medium,Small,28
16,536367,22622,BOX OF VINTAGE ALPHABET BLOCKS,2,2010-12-01 08:34:00,9.95,13047.0,UK,19.90,2010,12,1,8,1,Q4,Medium,Small,30
17,536367,21754,HOME BUILDING BLOCK WORD,3,2010-12-01 08:34:00,5.95,13047.0,UK,17.85,2010,12,1,8,1,Q4,Medium,Small,24
18,536367,21755,LOVE BUILDING BLOCK WORD,3,2010-12-01 08:34:00,5.95,13047.0,UK,17.85,2010,12,1,8,1,Q4,Medium,Small,24
19,536367,21777,RECIPE BOX WITH METAL HEART,4,2010-12-01 08:34:00,7.95,13047.0,UK,31.80,2010,12,1,8,1,Q4,Medium,Small,27
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
530050,581580,21351,CINAMMON & ORANGE WREATH,1,2011-12-09 12:20:00,6.75,12748.0,UK,6.75,2011,12,9,12,9,Q4,Low,Small,24
530053,581580,79321,CHILLI LIGHTS,2,2011-12-09 12:20:00,5.75,12748.0,UK,11.50,2011,12,9,12,9,Q4,Medium,Small,13
530073,581585,23356,LOVE HOT WATER BOTTLE,3,2011-12-09 12:31:00,5.95,15804.0,UK,17.85,2011,12,9,12,9,Q4,Medium,Small,21
530087,581586,21217,RED RETROSPOT ROUND CAKE TINS,24,2011-12-09 12:49:00,8.95,13113.0,UK,214.80,2011,12,9,12,9,Q4,High,Large,29


Task 46: Display transactions from France with Quantity greater than 20.

In [ ]:
onre[(onre['Country']=='France') & (onre['Quantity']>20)]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
26,536370,22728,ALARM CLOCK BAKELIKE PINK,24,2010-12-01 08:45:00,3.75,12583.0,France,90.00,2010,12,1,8,1,Q4,Medium,Large,25
27,536370,22727,ALARM CLOCK BAKELIKE RED,24,2010-12-01 08:45:00,3.75,12583.0,France,90.00,2010,12,1,8,1,Q4,Medium,Large,25
30,536370,21883,STARS GIFT TAPE,24,2010-12-01 08:45:00,0.65,12583.0,France,15.60,2010,12,1,8,1,Q4,Medium,Large,16
31,536370,10002,INFLATABLE POLITICAL GLOBE,48,2010-12-01 08:45:00,0.85,12583.0,France,40.80,2010,12,1,8,1,Q4,Medium,Large,27
32,536370,21791,VINTAGE HEADS AND TAILS CARD GAME,24,2010-12-01 08:45:00,1.25,12583.0,France,30.00,2010,12,1,8,1,Q4,Medium,Large,34
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
521245,581001,23084,RABBIT NIGHT LIGHT,24,2011-12-07 08:07:00,1.79,12583.0,France,42.96,2011,12,7,8,7,Q4,Medium,Large,18
521259,581001,22726,ALARM CLOCK BAKELIKE GREEN,24,2011-12-07 08:07:00,3.75,12583.0,France,90.00,2011,12,7,8,7,Q4,Medium,Large,26
523080,581171,22966,GINGERBREAD MAN COOKIE CUTTER,24,2011-12-07 15:02:00,1.25,12615.0,France,30.00,2011,12,7,15,7,Q4,Medium,Large,29
523091,581171,84077,WORLD WAR 2 GLIDERS ASSTD DESIGNS,48,2011-12-07 15:02:00,0.29,12615.0,France,13.92,2011,12,7,15,7,Q4,Medium,Large,33


Task 47: Display transactions from December 2010.

In [ ]:
onre[onre['InvoiceDate'].dt.strftime('%d/%m/%Y')=='01/12/2010']

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,UK,15.30,2010,12,1,8,1,Q4,Medium,Medium,34
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,19
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,UK,22.00,2010,12,1,8,1,Q4,Medium,Medium,30
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,35
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3067,536597,35271S,GOLD PRINT PAPER BAG,14,2010-12-01 17:35:00,0.19,18011.0,UK,2.66,2010,12,1,17,1,Q4,Low,Medium,20
3068,536597,21380,WOODEN HAPPY BIRTHDAY GARLAND,1,2010-12-01 17:35:00,2.95,18011.0,UK,2.95,2010,12,1,17,1,Q4,Low,Small,29
3069,536597,22909,SET OF 20 VINTAGE CHRISTMAS NAPKINS,1,2010-12-01 17:35:00,0.85,18011.0,UK,0.85,2010,12,1,17,1,Q4,Low,Small,35
3070,536597,21221,SET/4 BADGES CUTE CREATURES,5,2010-12-01 17:35:00,1.25,18011.0,UK,6.25,2010,12,1,17,1,Q4,Low,Medium,27


Task 48: Display transactions that occurred on weekends.

In [ ]:
onre[(onre['DayOfWeek']>5)]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
9998,537226,22811,SET OF 6 T-LIGHTS CACTI,6,2010-12-06 08:34:00,2.95,15987.0,UK,17.70,2010,12,6,8,6,Q4,Medium,Medium,24
9999,537226,21713,CITRONELLA CANDLE FLOWERPOT,8,2010-12-06 08:34:00,2.10,15987.0,UK,16.80,2010,12,6,8,6,Q4,Medium,Medium,27
10000,537226,22927,GREEN GIANT GARDEN THERMOMETER,2,2010-12-06 08:34:00,5.95,15987.0,UK,11.90,2010,12,6,8,6,Q4,Medium,Small,30
10001,537226,20802,SMALL GLASS SUNDAE DISH CLEAR,6,2010-12-06 08:34:00,1.65,15987.0,UK,9.90,2010,12,6,8,6,Q4,Low,Medium,29
10002,537226,22052,VINTAGE CARAVAN GIFT WRAP,25,2010-12-06 08:34:00,0.42,15987.0,UK,10.50,2010,12,6,8,6,Q4,Medium,Large,25
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
530099,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France,10.20,2011,12,9,12,9,Q4,Medium,Medium,27
530100,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France,12.60,2011,12,9,12,9,Q4,Medium,Medium,28
530101,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60,2011,12,9,12,9,Q4,Medium,Small,29
530102,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60,2011,12,9,12,9,Q4,Medium,Small,31


Task 49: Display the top 10 transactions by TotalPrice.

In [ ]:
onre.sort_values('TotalPrice',ascending=False).head(10)

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
528623,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,UK,168469.60,2011,12,9,9,9,Q4,High,Large,27
60250,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,UK,77183.60,2011,1,18,10,18,Q1,High,Large,30
217306,556444,22502,PICNIC BASKET WICKER 60 PIECES,60,2011-06-10 15:28:00,649.50,15098.0,UK,38970.00,2011,6,10,15,10,Q2,High,Large,30
14715,537632,AMAZONFEE,AMAZON FEE,1,2010-12-07 15:08:00,13541.33,NaN,UK,13541.33,2010,12,7,15,7,Q4,High,Small,10
292815,A563185,B,ADJUST BAD DEBT,1,2011-08-12 14:50:00,11062.06,NaN,UK,11062.06,2011,8,12,14,12,Q3,High,Small,15
169097,551697,POST,POSTAGE,1,2011-05-03 13:46:00,8142.75,16029.0,UK,8142.75,2011,5,3,13,3,Q2,High,Small,7
340186,567423,23243,SET OF TEA COFFEE SUGAR TINS PANTRY,1412,2011-09-20 11:05:00,5.06,17450.0,UK,7144.72,2011,9,20,11,20,Q3,High,Large,35
51432,540815,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-01-11 12:55:00,2.10,15749.0,UK,6539.40,2011,1,11,12,11,Q1,High,Large,34
156615,550461,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-04-18 13:20:00,2.10,15749.0,UK,6539.40,2011,4,18,13,18,Q2,High,Large,34
411682,573003,23084,RABBIT NIGHT LIGHT,2400,2011-10-27 12:11:00,2.08,14646.0,Netherlands,4992.00,2011,10,27,12,27,Q4,High,Large,18


Task 50: Display the bottom 15 transactions by TotalPrice.

In [ ]:
onre.sort_values('TotalPrice',ascending=False).tail(15)

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
377431,570238,16048,TEATIME ROUND PENCIL SHARPENER,1,2011-10-09 15:01:00,0.120,16332.0,UK,0.120,2011,10,9,15,9,Q4,Low,Small,31
434936,574831,16008,SMALL FOLDING SCISSOR(POINTED EDGE),1,2011-11-07 11:26:00,0.120,15522.0,UK,0.120,2011,11,7,11,7,Q4,Low,Small,35
448453,575856,16219,HOUSE SHAPE PENCIL SHARPENER,2,2011-11-11 12:28:00,0.060,16386.0,UK,0.120,2011,11,11,12,11,Q4,Low,Small,28
317269,565426,21403,BLUE EGG SPOON,1,2011-09-04 12:41:00,0.120,12851.0,UK,0.120,2011,9,4,12,4,Q3,Low,Small,15
269539,561036,85204,HANGING WOOD AND FELT FLOWER,1,2011-07-24 11:54:00,0.120,13137.0,UK,0.120,2011,7,24,11,24,Q3,Low,Small,28
250840,559509,17038,PORCELAIN BUDAH INCENSE HOLDER,1,2011-07-08 15:25:00,0.100,15920.0,UK,0.100,2011,7,8,15,8,Q3,Low,Small,30
486634,578479,17038,PORCELAIN BUDAH INCENSE HOLDER,1,2011-11-24 12:48:00,0.100,15858.0,UK,0.100,2011,11,24,12,24,Q4,Low,Small,30
235542,558198,17038,PORCELAIN BUDAH INCENSE HOLDER,1,2011-06-27 12:49:00,0.100,14104.0,UK,0.100,2011,6,27,12,27,Q2,Low,Small,30
278235,561894,M,MANUAL,1,2011-07-31 15:00:00,0.080,14456.0,UK,0.080,2011,7,31,15,31,Q3,Low,Small,6
489107,578757,16216,LETTER SHAPE PENCIL SHARPENER,1,2011-11-25 11:41:00,0.060,12748.0,UK,0.060,2011,11,25,11,25,Q4,Low,Small,29


Task 51: Find all transactions from either UK, Germany, or France.

In [ ]:
onre[onre['Country'].isin(['UK','Germany','France'])]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,UK,15.30,2010,12,1,8,1,Q4,Medium,Medium,34
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,19
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,UK,22.00,2010,12,1,8,1,Q4,Medium,Medium,30
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,35
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
530099,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France,10.20,2011,12,9,12,9,Q4,Medium,Medium,27
530100,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France,12.60,2011,12,9,12,9,Q4,Medium,Medium,28
530101,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60,2011,12,9,12,9,Q4,Medium,Small,29
530102,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60,2011,12,9,12,9,Q4,Medium,Small,31


Task 52: Find transactions where TotalPrice is above the 95th percentile.

In [ ]:
onre[onre['TotalPrice']>95.00]

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
46,536371,22086,PAPER CHAIN KIT 50'S CHRISTMAS,80,2010-12-01 09:00:00,2.55,13748.0,UK,204.00,2010,12,1,9,1,Q4,High,Large,31
65,536374,21258,VICTORIAN SEWING BOX LARGE,32,2010-12-01 09:09:00,10.95,15100.0,UK,350.40,2010,12,1,9,1,Q4,High,Large,26
82,536376,22114,HOT WATER BOTTLE TEA AND SYMPATHY,48,2010-12-01 09:32:00,3.45,15291.0,UK,165.60,2010,12,1,9,1,Q4,High,Large,33
83,536376,21733,RED HANGING HEART T-LIGHT HOLDER,64,2010-12-01 09:32:00,2.55,15291.0,UK,163.20,2010,12,1,9,1,Q4,High,Large,32
106,536381,22139,RETROSPOT TEA SET CERAMIC 11 PC,23,2010-12-01 09:41:00,4.25,15311.0,UK,97.75,2010,12,1,9,1,Q4,Medium,Large,32
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529712,581498,79321,CHILLI LIGHTS,10,2011-12-09 10:26:00,12.46,NaN,UK,124.60,2011,12,9,10,9,Q4,High,Medium,13
529739,581498,DOT,DOTCOM POSTAGE,1,2011-12-09 10:26:00,1714.17,NaN,UK,1714.17,2011,12,9,10,9,Q4,High,Small,14
529900,581566,23404,HOME SWEET HOME BLACKBOARD,144,2011-12-09 11:50:00,3.26,18102.0,UK,469.44,2011,12,9,11,9,Q4,High,Large,26
529946,581572,23328,SET 6 SCHOOL MILK BOTTLES IN CRATE,48,2011-12-09 12:08:00,3.39,16705.0,UK,162.72,2011,12,9,12,9,Q4,High,Large,34


#PART G: Sorting Operations

Task 53: Sort the entire dataset by InvoiceDate in ascending order

In [ ]:
onre.sort_values('InvoiceDate',ascending=True)

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,UK,15.30,2010,12,1,8,1,Q4,Medium,Medium,34
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,19
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,UK,22.00,2010,12,1,8,1,Q4,Medium,Medium,30
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,35
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,UK,20.34,2010,12,1,8,1,Q4,Medium,Medium,30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
530091,581587,22555,PLASTERS IN TIN STRONGMAN,12,2011-12-09 12:50:00,1.65,12680.0,France,19.80,2011,12,9,12,9,Q4,Medium,Medium,25
530090,581587,22556,PLASTERS IN TIN CIRCUS PARADE,12,2011-12-09 12:50:00,1.65,12680.0,France,19.80,2011,12,9,12,9,Q4,Medium,Medium,30
530102,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60,2011,12,9,12,9,Q4,Medium,Small,31
530095,581587,22730,ALARM CLOCK BAKELIKE IVORY,4,2011-12-09 12:50:00,3.75,12680.0,France,15.00,2011,12,9,12,9,Q4,Medium,Small,26


Task 54: Sort the dataset by TotalPrice in descending order.

In [ ]:
onre.sort_values('TotalPrice',ascending=False)

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
528623,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.080,16446.0,UK,168469.600,2011,12,9,9,9,Q4,High,Large,27
60250,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.040,12346.0,UK,77183.600,2011,1,18,10,18,Q1,High,Large,30
217306,556444,22502,PICNIC BASKET WICKER 60 PIECES,60,2011-06-10 15:28:00,649.500,15098.0,UK,38970.000,2011,6,10,15,10,Q2,High,Large,30
14715,537632,AMAZONFEE,AMAZON FEE,1,2010-12-07 15:08:00,13541.330,NaN,UK,13541.330,2010,12,7,15,7,Q4,High,Small,10
292815,A563185,B,ADJUST BAD DEBT,1,2011-08-12 14:50:00,11062.060,NaN,UK,11062.060,2011,8,12,14,12,Q3,High,Small,15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
414020,573174,16218,CARTOON PENCIL SHARPENERS,1,2011-10-28 10:25:00,0.060,NaN,UK,0.060,2011,10,28,10,28,Q4,Low,Small,26
272367,561226,PADS,PADS TO MATCH ALL CUSHIONS,1,2011-07-26 10:13:00,0.001,15618.0,UK,0.001,2011,7,26,10,26,Q3,Low,Small,26
153304,550193,PADS,PADS TO MATCH ALL CUSHIONS,1,2011-04-15 09:27:00,0.001,13952.0,UK,0.001,2011,4,15,9,15,Q2,Low,Small,26
353323,568375,BANK CHARGES,BANK CHARGES,1,2011-09-26 17:01:00,0.001,13405.0,UK,0.001,2011,9,26,17,26,Q3,Low,Small,12


Task 55: Sort by Country (ascending) and then by TotalPrice (descending).

In [ ]:
onre.sort_values(['Country','TotalPrice'],ascending=[True,False])

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
457455,576394,23084,RABBIT NIGHT LIGHT,960,2011-11-15 10:32:00,1.79,12415.0,Australia,1718.40,2011,11,15,10,15,Q4,High,Large,18
297617,563614,22940,FELTCRAFT CHRISTMAS FAIRY,336,2011-08-18 08:51:00,3.75,12415.0,Australia,1260.00,2011,8,18,8,18,Q3,High,Large,25
105512,545475,21217,RED RETROSPOT ROUND CAKE TINS,120,2011-03-03 10:59:00,8.95,12415.0,Australia,1074.00,2011,3,3,10,3,Q1,High,Large,29
297687,563614,23314,VINTAGE CHRISTMAS TABLECLOTH,100,2011-08-18 08:51:00,10.39,12415.0,Australia,1039.00,2011,8,18,8,18,Q3,High,Large,28
44433,540267,22720,SET OF 3 CAKE TINS PANTRY DESIGN,240,2011-01-06 11:12:00,4.25,12415.0,Australia,1020.00,2011,1,6,11,6,Q1,High,Large,33
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
315726,565303,22541,MINI JIGSAW LEAP FROG,1,2011-09-02 12:17:00,0.42,NaN,Unspecified,0.42,2011,9,2,12,2,Q3,Low,Small,21
315727,565303,22540,MINI JIGSAW CIRCUS PARADE,1,2011-09-02 12:17:00,0.42,NaN,Unspecified,0.42,2011,9,2,12,2,Q3,Low,Small,26
275978,561658,47348A,FUSCHIA VOILE POINTY SHOE DEC,1,2011-07-28 16:06:00,0.39,12743.0,Unspecified,0.39,2011,7,28,16,28,Q3,Low,Small,29
276029,561661,22396,MAGNETS PACK OF 4 RETRO PHOTO,1,2011-07-28 16:21:00,0.39,12743.0,Unspecified,0.39,2011,7,28,16,28,Q3,Low,Small,29


Task 56: Find the 20 highest-value transactions.

In [ ]:
onre.sort_values('TotalPrice',ascending=False).head(20)

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
528623,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,UK,168469.60,2011,12,9,9,9,Q4,High,Large,27
60250,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,UK,77183.60,2011,1,18,10,18,Q1,High,Large,30
217306,556444,22502,PICNIC BASKET WICKER 60 PIECES,60,2011-06-10 15:28:00,649.50,15098.0,UK,38970.00,2011,6,10,15,10,Q2,High,Large,30
14715,537632,AMAZONFEE,AMAZON FEE,1,2010-12-07 15:08:00,13541.33,NaN,UK,13541.33,2010,12,7,15,7,Q4,High,Small,10
292815,A563185,B,ADJUST BAD DEBT,1,2011-08-12 14:50:00,11062.06,NaN,UK,11062.06,2011,8,12,14,12,Q3,High,Small,15
169097,551697,POST,POSTAGE,1,2011-05-03 13:46:00,8142.75,16029.0,UK,8142.75,2011,5,3,13,3,Q2,High,Small,7
340186,567423,23243,SET OF TEA COFFEE SUGAR TINS PANTRY,1412,2011-09-20 11:05:00,5.06,17450.0,UK,7144.72,2011,9,20,11,20,Q3,High,Large,35
51432,540815,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-01-11 12:55:00,2.10,15749.0,UK,6539.40,2011,1,11,12,11,Q1,High,Large,34
156615,550461,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-04-18 13:20:00,2.10,15749.0,UK,6539.40,2011,4,18,13,18,Q2,High,Large,34
411682,573003,23084,RABBIT NIGHT LIGHT,2400,2011-10-27 12:11:00,2.08,14646.0,Netherlands,4992.00,2011,10,27,12,27,Q4,High,Large,18


Task 57: Find the 20 lowest-value transactions (excluding returns).

In [ ]:
onre.sort_values('TotalPrice',ascending=False).tail(20)

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName
385667,570983,47518F,ICON PLACEMAT POP ART ELVIS,1,2011-10-13 11:57:00,0.120,14963.0,UK,0.120,2011,10,13,11,13,Q4,Low,Small,27
432543,574682,20668,DISCO BALL CHRISTMAS DECORATION,1,2011-11-06 12:44:00,0.120,14963.0,UK,0.120,2011,11,6,12,6,Q4,Low,Small,31
398049,571909,16048,TEATIME ROUND PENCIL SHARPENER,1,2011-10-19 15:08:00,0.120,15006.0,UK,0.120,2011,10,19,15,19,Q4,Low,Small,31
303272,564191,85204,HANGING WOOD AND FELT FLOWER,1,2011-08-23 16:16:00,0.120,17315.0,UK,0.120,2011,8,23,16,23,Q3,Low,Small,28
525759,581310,22422,TOOTHPASTE TUBE PEN,1,2011-12-08 11:43:00,0.120,16442.0,UK,0.120,2011,12,8,11,8,Q4,Low,Small,19
377431,570238,16048,TEATIME ROUND PENCIL SHARPENER,1,2011-10-09 15:01:00,0.120,16332.0,UK,0.120,2011,10,9,15,9,Q4,Low,Small,31
434936,574831,16008,SMALL FOLDING SCISSOR(POINTED EDGE),1,2011-11-07 11:26:00,0.120,15522.0,UK,0.120,2011,11,7,11,7,Q4,Low,Small,35
448453,575856,16219,HOUSE SHAPE PENCIL SHARPENER,2,2011-11-11 12:28:00,0.060,16386.0,UK,0.120,2011,11,11,12,11,Q4,Low,Small,28
317269,565426,21403,BLUE EGG SPOON,1,2011-09-04 12:41:00,0.120,12851.0,UK,0.120,2011,9,4,12,4,Q3,Low,Small,15
269539,561036,85204,HANGING WOOD AND FELT FLOWER,1,2011-07-24 11:54:00,0.120,13137.0,UK,0.120,2011,7,24,11,24,Q3,Low,Small,28


#PART H: Aggregation and Summary Statistics

Task 58: Calculate the total revenue generated across all transactions.

In [ ]:
onre['TotalPrice'].sum()

np.float64(10666684.544)

Task 59: Calculate the average transaction value.

In [ ]:
onre['TotalPrice'].mean()

np.float64(20.121871451639677)

Task 60: Find the maximum and minimum transaction values.

In [ ]:
onre['TotalPrice'].min()

0.001

In [ ]:
onre['TotalPrice'].max()

168469.6

Task 61: Calculate the total number of unique customers.

In [ ]:
onre['CustomerID'].nunique()

4338

Task 62: Calculate the total number of unique products sold.

In [ ]:
onre['ProductName'].nunique()

4026

Task 63: Find out how many unique countries are represented in the data.

In [ ]:
onre['Country'].nunique()

38

Task 64: Calculate the average number of items per transaction.

In [ ]:
onre['Quantity'].mean()

np.float64(10.542037034242338)

Task 65: Find which product appears most frequently in transactions.

In [ ]:
onre['ProductName'].mode()

,ProductName
0,WHITE HANGING HEART T-LIGHT HOLDER


#PART I: Country-wise Analysis

Task 66: Calculate total revenue for each country.

In [ ]:
onre.groupby('Country')['TotalPrice'].sum()

,TotalPrice
Country,
Australia,138521.310
Austria,10198.680
Bahrain,754.140
Belgium,41196.340
Brazil,1143.600
Canada,3666.380
Channel Islands,20450.440
Cyprus,13590.380
Czech Republic,826.740


Task 67: Find the top 10 countries by total revenue.

In [ ]:
onre.groupby('Country')['TotalPrice'].sum().sort_values(ascending=False).head(10)

,TotalPrice
Country,
UK,9025222.084
Netherlands,285446.340
EIRE,283453.960
Germany,228867.140
France,209715.110
Australia,138521.310
Spain,61577.110
Switzerland,57089.900
Belgium,41196.340


Task 68: Calculate the number of transactions per country.

In [ ]:
onre.groupby('Country')['InvoiceNo'].count()

,InvoiceNo
Country,
Australia,1182
Austria,398
Bahrain,18
Belgium,2031
Brazil,32
Canada,151
Channel Islands,748
Cyprus,614
Czech Republic,25


Task 69: Calculate the average transaction value for each country.

In [ ]:
onre.groupby('Country')['TotalPrice'].mean()

,TotalPrice
Country,
Australia,117.192310
Austria,25.624824
Bahrain,41.896667
Belgium,20.283772
Brazil,35.737500
Canada,24.280662
Channel Islands,27.340160
Cyprus,22.134169
Czech Republic,33.069600


Task 70: Find which country has the highest average order value.

In [ ]:
onre.groupby('Country')['TotalPrice'].mean().sort_values(ascending=False).head(1)

,TotalPrice
Country,
Netherlands,121.003111


Task 71: Calculate the total quantity of products sold to each country.

In [ ]:
onre.groupby('Country')['Quantity'].sum()

,Quantity
Country,
Australia,83901
Austria,4881
Bahrain,314
Belgium,23237
Brazil,356
Canada,2763
Channel Islands,9491
Cyprus,6361
Czech Republic,671


Task 72: Find the country with the most unique customers.

In [ ]:
onre.groupby('Country')['CustomerID'].nunique().sort_values(ascending=False).head(1)

,CustomerID
Country,
UK,3920


#PART J: Time-based Analysis

Task 73: Calculate total revenue for each month.

In [ ]:
onre.groupby(onre['InvoiceDate'].dt.month)['TotalPrice'].sum()

,TotalPrice
InvoiceDate,
1,691364.560
2,523631.890
3,717639.360
4,537808.621
5,770536.020
6,761739.900
7,719221.191
8,759138.380
9,1058590.172


Task 74: Find which month had the highest sales.

In [ ]:
onre.groupby(onre['InvoiceDate'].dt.month)['TotalPrice'].sum().sort_values(ascending=False).head(1)

,TotalPrice
InvoiceDate,
11,1509496.33


Task 75: Calculate revenue for each quarter.

In [ ]:
onre.groupby(onre['InvoiceDate'].dt.quarter)['TotalPrice'].sum()

,TotalPrice
InvoiceDate,
1,1932635.810
2,2070084.541
3,2536949.743
4,4127014.450


Task 76: Calculate the number of transactions per day of the week.

In [ ]:
onre.groupby(onre['InvoiceDate'].dt.day_of_week)['InvoiceNo'].count()

,InvoiceNo
InvoiceDate,
0,93136
1,99460
2,92315
3,101007
4,80281
6,63905


Task 77: Find which day of the week has the highest average revenue.

In [ ]:
onre.groupby(onre['InvoiceDate'].dt.day_of_week)['TotalPrice'].mean().sort_values(ascending=False).head(1)

,TotalPrice
InvoiceDate,
4,22.923733


Task 78: Calculate hourly sales patterns (revenue by hour of day).

In [ ]:
onre.groupby(onre['InvoiceDate'].dt.hour)['TotalPrice'].sum()

,TotalPrice
InvoiceDate,
6,4.250
7,31059.210
8,283868.520
9,990267.821
10,1446742.701
11,1239954.440
12,1444245.880
13,1265736.300
14,1181812.371


Task 79: Compare sales between 2010 and 2011.

In [ ]:
sby = onre.groupby(onre['InvoiceDate'].dt.year)['TotalPrice'].sum()
sby

,TotalPrice
InvoiceDate,
2010,823746.140
2011,9842938.404


Task 80: Calculate month-over-month revenue growth rate.

In [ ]:
onre.groupby(onre['InvoiceDate'].dt.month)['TotalPrice'].sum().pct_change()

,TotalPrice
InvoiceDate,
1,NaN
2,-0.242611
3,0.370504
4,-0.250587
5,0.432733
6,-0.011416
7,-0.055818
8,0.055501
9,0.394463


#PART K: Customer Analysis

Task 81: Calculate total spending for each customer.

In [ ]:
onre.groupby('CustomerID')['TotalPrice'].sum()

,TotalPrice
CustomerID,
12346.0,77183.60
12347.0,4310.00
12348.0,1797.24
12349.0,1757.55
12350.0,334.40
...,...
18280.0,180.60
18281.0,80.82
18282.0,178.05


Task 82: Find the top 20 customers by total spending.

In [ ]:
onre.groupby(['CustomerID'])['TotalPrice'].sum().sort_values(ascending=False).head(20)

,TotalPrice
CustomerID,
14646.0,280206.02
18102.0,259657.30
17450.0,194550.79
16446.0,168472.50
14911.0,143825.06
12415.0,124914.53
14156.0,117379.63
17511.0,91062.38
16029.0,81024.84


Task 83: Calculate the number of transactions per customer.

In [ ]:
onre.groupby(['CustomerID'])['InvoiceNo'].count()

,InvoiceNo
CustomerID,
12346.0,1
12347.0,182
12348.0,31
12349.0,73
12350.0,17
...,...
18280.0,10
18281.0,7
18282.0,12


Task 84: Find the average order value for each customer.

In [ ]:
onre.groupby(['CustomerID'])['TotalPrice'].mean()

,TotalPrice
CustomerID,
12346.0,77183.600000
12347.0,23.681319
12348.0,57.975484
12349.0,24.076027
12350.0,19.670588
...,...
18280.0,18.060000
18281.0,11.545714
18282.0,14.837500


Task 85: Identify customers who have made only one purchase.

In [ ]:
onep=onre.groupby(['CustomerID'])['InvoiceNo'].count()
onep[onep==1]

,InvoiceNo
CustomerID,
12346.0,1
13017.0,1
13099.0,1
13106.0,1
13120.0,1
...,...
18133.0,1
18174.0,1
18184.0,1


Task 86: Find customers who spent more than £10,000 in total.

In [ ]:
tp = onre.groupby(['CustomerID'])['TotalPrice'].sum()
tp[tp>10000]

,TotalPrice
CustomerID,
12346.0,77183.60
12409.0,11072.67
12415.0,124914.53
12433.0,13375.87
12471.0,19824.05
...,...
17841.0,40991.57
17857.0,26879.04
17865.0,10526.32


Task 87: Calculate the average number of days between purchases for each customer.

In [ ]:
onre = onre.sort_values(['CustomerID','InvoiceDate'])

In [ ]:
onre['Diff'] = onre.groupby('CustomerID')['InvoiceDate'].diff().dt.days

In [ ]:
avg = onre.groupby('CustomerID')['Diff'].mean()
avg

,Diff
CustomerID,
12346.0,NaN
12347.0,2.000000
12348.0,9.400000
12349.0,0.000000
12350.0,0.000000
...,...
18280.0,0.000000
18281.0,0.000000
18282.0,10.727273


#PART L: Product Analysis

Task 88: Find the top 20 products by revenue.

In [ ]:
onre.groupby('ProductName')['TotalPrice'].sum().sort_values(ascending=False).head(20)

,TotalPrice
ProductName,
DOTCOM POSTAGE,206248.77
REGENCY CAKESTAND 3 TIER,174484.74
"PAPER CRAFT , LITTLE BIRDIE",168469.60
WHITE HANGING HEART T-LIGHT HOLDER,106292.77
PARTY BUNTING,99504.33
JUMBO BAG RED RETROSPOT,94340.05
MEDIUM CERAMIC TOP STORAGE JAR,81700.92
MANUAL,78112.82
POSTAGE,78101.88


Task 89: Find the top 20 products by quantity sold.

In [ ]:
onre.groupby('Quantity')['ProductName'].sum().sort_values(ascending=False).head(20)

,ProductName
Quantity,
25,WRAP RED VINTAGE DOILYWRAP RED APPLES WRAP WED...
75,WRAP RED APPLES WRAP RED VINTAGE DOILYWRAP SUK...
400,WRAP ENGLISH ROSE WRAP RED APPLES LUNCH BOX I ...
200,WRAP DOILEY DESIGNWRAP ALPHABET DESIGNSET 2 PA...
1200,WORLD WAR 2 GLIDERS ASSTD DESIGNSRAIN PONCHO R...
1008,WORLD WAR 2 GLIDERS ASSTD DESIGNSPACK OF 12 LO...
2592,WORLD WAR 2 GLIDERS ASSTD DESIGNS
4800,WORLD WAR 2 GLIDERS ASSTD DESIGNS
11,WOODLAND DESIGN COTTON TOTE BAGPOSTAGEPOSTAGE...


Task 90: Calculate the average price for each product.

In [ ]:
onre.groupby('ProductName')['TotalPrice'].mean()

,TotalPrice
ProductName,
4 PURPLE FLOCK DINNER CANDLES,7.092683
50'S CHRISTMAS GIFT BAG LARGE,18.167674
DOLLY GIRL BEAKER,16.152235
I LOVE LONDON MINI BACKPACK,18.501932
I LOVE LONDON MINI RUCKSACK,4.150000
...,...
ZINC T-LIGHT HOLDER STARS SMALL,15.264836
ZINC TOP 2 DOOR WOODEN SHELF,22.506667
ZINC WILLIE WINKIE CANDLE STICK,9.976287


Task 91: Find products that have been sold to more than 10 different countries.

In [ ]:
cnty = onre.groupby('ProductName')['Country'].nunique()
cnty[cnty>10]

,Country
ProductName,
DOLLY GIRL BEAKER,11
SPACEBOY BABY GIFT SET,13
10 COLOUR SPACEBOY PEN,13
12 PENCIL SMALL TUBE WOODLAND,12
12 PENCILS SMALL TUBE RED RETROSPOT,12
...,...
WRAP PAISLEY PARK,14
WRAP RED APPLES,14
WRAP RED VINTAGE DOILY,12


Task 92: Identify the most expensive and cheapest products.

In [ ]:
onre.groupby('ProductName')['TotalPrice'].max().sort_values(ascending=False).head(1)

,TotalPrice
ProductName,
"PAPER CRAFT , LITTLE BIRDIE",168469.6


#PART M: Advanced Grouping

Task 93: Calculate total revenue by Country and Year.

In [ ]:
onre.groupby(['Country','Year'])['TotalPrice'].sum()

Country               Year
Australia             2010       1032.850
                      2011     137488.460
Austria               2010        277.200
                      2011       9921.480
Bahrain               2010        205.740
                      2011        548.400
Belgium               2010       1809.910
                      2011      39386.430
Brazil                2011       1143.600
Canada                2011       3666.380
Channel Islands       2010        363.530
                      2011      20086.910
Cyprus                2010       1590.820
                      2011      11999.560
Czech Republic        2011        826.740
Denmark               2010       1281.500
                      2011      17673.840
EIRE                  2010      10033.260
                      2011     273420.700
European Community    2011       1300.250
Finland               2010        892.800
                      2011      21653.280
France                2010       9616.310
                      2011     200098.800
Germany               2010      15241.140
                      2011     213626.000
Greece                2011       4760.520
Hong Kong             2011      15691.800
Iceland               2010        711.790
                      2011       3598.210
Israel                2011       8135.260
Italy                 2010        811.500
                      2011      16671.740
Japan                 2010       7705.070
                      2011      29711.300
Lebanon               2011       1693.880
Lithuania             2010       1661.060
Malta                 2011       2725.590
Netherlands           2010       8784.480
                      2011     276661.860
Norway                2010       3787.120
                      2011      32378.320
Poland                2010        248.160
                      2011       7086.490
Portugal              2010       2439.970
                      2011      31307.130
RSA                   2011       1002.310
Saudi Arabia          2011        145.920
Singapore             2011      21279.290
Spain                 2010       1843.730
                      2011      59733.380
Sweden                2010       3834.300
                      2011      34544.030
Switzerland           2010       1304.920
                      2011      55784.980
UK                    2010     748268.980
                      2011    8276953.104
USA                   2011       3580.390
United Arab Emirates  2011       1902.280
Unspecified           2011       4749.790
Name: TotalPrice, dtype: float64

Task 94: Calculate average transaction value by Country and Quarter.

In [ ]:
onre.groupby(['Country','Quarter'])['TotalPrice'].mean()

Country               Quarter
Australia             Q1         126.349136
                      Q2         130.685743
                      Q3          88.479103
                      Q4         135.968449
Austria               Q1          57.089231
                                    ...    
United Arab Emirates  Q1          29.641333
                      Q3          26.658947
Unspecified           Q2          18.563253
                      Q3           7.236419
                      Q4          18.221698
Name: TotalPrice, Length: 124, dtype: float64

Task 95: Find the number of transactions by Country and Month.

In [ ]:
onre.groupby(['Country','Month'])['InvoiceNo'].count()

Country      Month
Australia    1        127
             2         89
             3        108
             4         18
             5        116
                     ... 
Unspecified  6          9
             7        221
             8         23
             9         66
             11        53
Name: InvoiceNo, Length: 286, dtype: int64

Task 96: Calculate total revenue by Day of Week and Hour.

In [ ]:
onre.groupby(['DayOfWeek','Hour'])['TotalPrice'].sum()

DayOfWeek  Hour
1          7         326.90
           8        9311.84
           9       25497.36
           10      38438.44
           11      38179.01
                     ...   
31         15      37642.98
           16      10805.31
           17       3875.58
           18        551.65
           19        984.90
Name: TotalPrice, Length: 398, dtype: float64

Task 97: For each country, calculate the sum, mean, maximum, and minimum TotalPrice.

In [ ]:
onre.groupby('Country')['TotalPrice'].agg(['sum','mean','max','min'])

,sum,mean,max,min
Country,,,,
Australia,138521.310,117.192310,1718.40,0.420
Austria,10198.680,25.624824,360.00,2.880
Bahrain,754.140,41.896667,205.74,11.600
Belgium,41196.340,20.283772,165.00,2.500
Brazil,1143.600,35.737500,175.20,15.000
Canada,3666.380,24.280662,550.94,2.500
Channel Islands,20450.440,27.340160,408.00,4.560
Cyprus,13590.380,22.134169,320.69,0.830
Czech Republic,826.740,33.069600,70.80,10.080


Task 98: For each customer, calculate their total spending, number of orders, and average order value.

In [ ]:
onre.groupby('CustomerID')['TotalPrice'].agg(['sum','count','mean'])

,sum,count,mean
CustomerID,,,
12346.0,77183.60,1,77183.600000
12347.0,4310.00,182,23.681319
12348.0,1797.24,31,57.975484
12349.0,1757.55,73,24.076027
12350.0,334.40,17,19.670588
...,...,...,...
18280.0,180.60,10,18.060000
18281.0,80.82,7,11.545714
18282.0,178.05,12,14.837500


Task 99: For each product, calculate total quantity sold and total revenue generated.

In [ ]:
onre.groupby('ProductName')['TotalPrice'].sum()

,TotalPrice
ProductName,
4 PURPLE FLOCK DINNER CANDLES,290.80
50'S CHRISTMAS GIFT BAG LARGE,2343.63
DOLLY GIRL BEAKER,2891.25
I LOVE LONDON MINI BACKPACK,1628.17
I LOVE LONDON MINI RUCKSACK,4.15
...,...
ZINC T-LIGHT HOLDER STARS SMALL,4197.83
ZINC TOP 2 DOOR WOODEN SHELF,270.08
ZINC WILLIE WINKIE CANDLE STICK,2713.55


In [ ]:
onre.groupby('ProductName')['Quantity'].sum()

,Quantity
ProductName,
4 PURPLE FLOCK DINNER CANDLES,144
50'S CHRISTMAS GIFT BAG LARGE,1915
DOLLY GIRL BEAKER,2455
I LOVE LONDON MINI BACKPACK,389
I LOVE LONDON MINI RUCKSACK,1
...,...
ZINC T-LIGHT HOLDER STARS SMALL,5089
ZINC TOP 2 DOOR WOODEN SHELF,13
ZINC WILLIE WINKIE CANDLE STICK,2934


#PART N: Creating Summary Reports

Task 100: Create a summary table showing monthly sales for each country (countries as rows, months as columns).

In [ ]:
onre.pivot_table(index='Country',columns=onre['InvoiceDate'].dt.month,values='TotalPrice',aggfunc='sum')

InvoiceDate,1,2,3,4,5,6,7,8,9,10,11,12
Country,,,,,,,,,,,,
Australia,9017.71,14695.42,17223.99,771.600,13638.41,25187.77,4964.380,22489.20,5106.730,17150.53,7242.72,1032.85
Austria,NaN,518.36,1708.12,680.780,1249.43,NaN,1191.950,1516.08,NaN,1043.78,1329.78,960.40
Bahrain,NaN,NaN,NaN,NaN,548.40,NaN,NaN,NaN,NaN,NaN,NaN,205.74
Belgium,1200.20,2181.07,3351.98,1989.480,2732.40,4274.82,2475.570,3554.02,4208.020,5685.38,6315.76,3227.64
Brazil,NaN,NaN,NaN,1143.600,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Canada,NaN,NaN,140.54,NaN,534.24,1171.46,1768.580,51.56,NaN,NaN,NaN,NaN
Channel Islands,675.58,1784.71,3509.33,293.000,1207.24,2060.03,NaN,4896.78,1323.750,2623.32,1514.77,561.93
Cyprus,547.50,4334.24,938.39,NaN,NaN,1109.32,NaN,NaN,196.350,4412.87,460.89,1590.82
Czech Republic,NaN,549.26,NaN,NaN,NaN,NaN,NaN,NaN,NaN,277.48,NaN,NaN


Task 101: Create a summary showing average quantity purchased by country and quarter.

In [ ]:
onre.pivot_table(index='Country',columns='Quarter',values='Quantity',aggfunc='mean')

Quarter,Q1,Q2,Q3,Q4
Country,,,,
Australia,76.429012,83.184818,59.692935,63.989305
Austria,33.948718,10.220779,9.531469,10.122302
Bahrain,NaN,15.294118,NaN,54.000000
Belgium,10.943452,10.663286,11.298419,12.336207
Brazil,NaN,11.125000,NaN,NaN
Canada,8.800000,27.517241,13.000000,NaN
Channel Islands,14.855856,9.330645,13.526829,11.487310
Cyprus,15.181818,7.574468,33.000000,7.606936
Czech Republic,23.066667,NaN,NaN,32.500000


Task 102: Create a table showing the count of transactions in each RevenueCategory by Country.

In [ ]:
onre.pivot_table(index='Country',columns='RevenueCategory',values='InvoiceNo',aggfunc='count')

RevenueCategory,High,Low,Medium
Country,,,
Australia,445.0,82.0,655.0
Austria,13.0,37.0,348.0
Bahrain,2.0,NaN,16.0
Belgium,16.0,209.0,1806.0
Brazil,1.0,NaN,31.0
Canada,2.0,19.0,130.0
Channel Islands,22.0,21.0,705.0
Cyprus,14.0,150.0,450.0
Czech Republic,NaN,NaN,25.0


#PART O: Data Transformation

Task 103: Add a column to each row showing what percentage of the monthly total revenue that transaction represents.

In [ ]:
trans = onre.groupby(onre['InvoiceDate'].dt.month)['TotalPrice'].transform('sum')

In [ ]:
onre['trans'] = onre['TotalPrice']/trans
onre['trans']

,trans
60250,0.111640
14637,0.000017
14638,0.000012
14639,0.000027
14640,0.000016
...,...
529735,0.000014
529736,0.000011
529737,0.000003
529738,0.000007


Task 104: Add a column showing each customer's ranking based on their total spending

In [ ]:
onre['rank'] = onre.groupby('CustomerID')['TotalPrice'].rank(ascending=False)
onre['rank']

,rank
60250,1.0
14637,46.0
14638,97.5
14639,17.0
14640,52.0
...,...
529735,NaN
529736,NaN
529737,NaN
529738,NaN


Task 105: Add a column showing the cumulative revenue over time.

In [ ]:
onre['Cumulatives'] = onre.groupby('InvoiceDate')['TotalPrice'].cumsum()
onre['Cumulatives']

,Cumulatives
60250,77183.60
14637,25.20
14638,42.20
14639,81.20
14640,104.60
...,...
529735,3918.45
529736,3934.97
529737,3939.93
529738,3950.72


#PART P: Combining Analysis

Task 106: Split the dataset into two parts: 2010 data and 2011 data. Then combine them back together vertically.

In [ ]:
onre2010 = onre[onre['InvoiceDate'].dt.year==2010]
onre2010

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,...,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName,Diff,trans,rank,Cumulatives
14637,537626,85116,BLACK CANDELABRA T-LIGHT HOLDER,12,2010-12-07 14:57:00,2.10,12347.0,Iceland,25.20,2010,...,14,7,Q4,Medium,Medium,31,NaN,1.723031e-05,46.0,25.20
14638,537626,22375,AIRLINE BAG VINTAGE JET SET BROWN,4,2010-12-07 14:57:00,4.25,12347.0,Iceland,17.00,2010,...,14,7,Q4,Medium,Small,33,0.0,1.162362e-05,97.5,42.20
14639,537626,71477,COLOUR GLASS. STAR T-LIGHT HOLDER,12,2010-12-07 14:57:00,3.25,12347.0,Iceland,39.00,2010,...,14,7,Q4,Medium,Medium,33,0.0,2.666596e-05,17.0,81.20
14640,537626,22492,MINI PAINT SET VINTAGE,36,2010-12-07 14:57:00,0.65,12347.0,Iceland,23.40,2010,...,14,7,Q4,Medium,Large,23,0.0,1.599958e-05,52.0,104.60
14641,537626,22771,CLEAR DRAWER KNOB ACRYLIC EDWARDIAN,12,2010-12-07 14:57:00,1.25,12347.0,Iceland,15.00,2010,...,14,7,Q4,Medium,Medium,35,0.0,1.025614e-05,119.5,119.60
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
41475,539991,21618,4 WILDFLOWER BOTANICAL CANDLES,1,2010-12-23 16:49:00,1.25,NaN,UK,1.25,2010,...,16,23,Q4,Low,Small,30,NaN,8.546782e-07,NaN,72.90
41476,539991,72741,GRAND CHOCOLATECANDLE,4,2010-12-23 16:49:00,1.45,NaN,UK,5.80,2010,...,16,23,Q4,Low,Small,21,NaN,3.965707e-06,NaN,78.70
41477,539992,21470,FLOWER VINE RAFFIA FOOD COVER,1,2010-12-23 17:41:00,3.75,NaN,UK,3.75,2010,...,17,23,Q4,Low,Small,29,NaN,2.564035e-06,NaN,3.75
41478,539992,22258,FELT FARM ANIMAL RABBIT,1,2010-12-23 17:41:00,1.25,NaN,UK,1.25,2010,...,17,23,Q4,Low,Small,23,NaN,8.546782e-07,NaN,5.00


In [ ]:
onre2011 = onre[onre['InvoiceDate'].dt.year==2011]
onre2011

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,...,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName,Diff,trans,rank,Cumulatives
60250,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,UK,77183.60,2011,...,10,18,Q1,High,Large,30,NaN,0.111640,1.0,77183.60
70665,542237,84625A,PINK NEW BAROQUECANDLESTICK CANDLE,24,2011-01-26 14:30:00,0.85,12347.0,Iceland,20.40,2011,...,14,26,Q1,Medium,Large,34,49.0,0.000030,63.0,20.40
70666,542237,84625C,BLUE NEW BAROQUE CANDLESTICK CANDLE,24,2011-01-26 14:30:00,0.85,12347.0,Iceland,20.40,2011,...,14,26,Q1,Medium,Large,35,0.0,0.000030,63.0,40.80
70667,542237,85116,BLACK CANDELABRA T-LIGHT HOLDER,6,2011-01-26 14:30:00,2.10,12347.0,Iceland,12.60,2011,...,14,26,Q1,Medium,Medium,31,0.0,0.000018,153.5,53.40
70668,542237,20719,WOODLAND CHARLOTTE BAG,10,2011-01-26 14:30:00,0.85,12347.0,Iceland,8.50,2011,...,14,26,Q1,Low,Medium,22,0.0,0.000012,172.5,61.90
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529735,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,NaN,UK,20.65,2011,...,10,9,Q4,Medium,Medium,23,NaN,0.000014,NaN,3918.45
529736,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,NaN,UK,16.52,2011,...,10,9,Q4,Medium,Small,30,NaN,0.000011,NaN,3934.97
529737,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,NaN,UK,4.96,2011,...,10,9,Q4,Low,Small,29,NaN,0.000003,NaN,3939.93
529738,581498,85174,S/4 CACTI CANDLES,1,2011-12-09 10:26:00,10.79,NaN,UK,10.79,2011,...,10,9,Q4,Medium,Small,17,NaN,0.000007,NaN,3950.72


In [ ]:
onrecon = pd.concat([onre2010,onre2011])
onrecon

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,...,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName,Diff,trans,rank,Cumulatives
14637,537626,85116,BLACK CANDELABRA T-LIGHT HOLDER,12,2010-12-07 14:57:00,2.10,12347.0,Iceland,25.20,2010,...,14,7,Q4,Medium,Medium,31,NaN,0.000017,46.0,25.20
14638,537626,22375,AIRLINE BAG VINTAGE JET SET BROWN,4,2010-12-07 14:57:00,4.25,12347.0,Iceland,17.00,2010,...,14,7,Q4,Medium,Small,33,0.0,0.000012,97.5,42.20
14639,537626,71477,COLOUR GLASS. STAR T-LIGHT HOLDER,12,2010-12-07 14:57:00,3.25,12347.0,Iceland,39.00,2010,...,14,7,Q4,Medium,Medium,33,0.0,0.000027,17.0,81.20
14640,537626,22492,MINI PAINT SET VINTAGE,36,2010-12-07 14:57:00,0.65,12347.0,Iceland,23.40,2010,...,14,7,Q4,Medium,Large,23,0.0,0.000016,52.0,104.60
14641,537626,22771,CLEAR DRAWER KNOB ACRYLIC EDWARDIAN,12,2010-12-07 14:57:00,1.25,12347.0,Iceland,15.00,2010,...,14,7,Q4,Medium,Medium,35,0.0,0.000010,119.5,119.60
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529735,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,NaN,UK,20.65,2011,...,10,9,Q4,Medium,Medium,23,NaN,0.000014,NaN,3918.45
529736,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,NaN,UK,16.52,2011,...,10,9,Q4,Medium,Small,30,NaN,0.000011,NaN,3934.97
529737,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,NaN,UK,4.96,2011,...,10,9,Q4,Low,Small,29,NaN,0.000003,NaN,3939.93
529738,581498,85174,S/4 CACTI CANDLES,1,2011-12-09 10:26:00,10.79,NaN,UK,10.79,2011,...,10,9,Q4,Medium,Small,17,NaN,0.000007,NaN,3950.72


Task 107: Create separate datasets for UK and non-UK transactions. Combine them back horizontally with appropriate labeling.

In [ ]:
uk = onre[onre['Country']=='UK']
uk

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,...,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName,Diff,trans,rank,Cumulatives
60250,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,UK,77183.60,2011,...,10,18,Q1,High,Large,30,NaN,0.111640,1.0,77183.60
9608,537215,85124C,GREEN JUICY FRUIT PHOTO FRAME,12,2010-12-05 15:38:00,2.55,12747.0,UK,30.60,2010,...,15,5,Q4,Medium,Medium,29,NaN,0.000021,37.0,30.60
9609,537215,85124B,BLUE JUICY FRUIT PHOTO FRAME,6,2010-12-05 15:38:00,2.55,12747.0,UK,15.30,2010,...,15,5,Q4,Medium,Medium,28,0.0,0.000010,90.0,45.90
9610,537215,84879,ASSORTED COLOUR BIRD ORNAMENT,16,2010-12-05 15:38:00,1.69,12747.0,UK,27.04,2010,...,15,5,Q4,Medium,Medium,29,0.0,0.000018,46.5,72.94
9611,537215,85062,PEARL CRYSTAL PUMPKIN T-LIGHT HLDR,24,2010-12-05 15:38:00,1.65,12747.0,UK,39.60,2010,...,15,5,Q4,Medium,Large,34,0.0,0.000027,23.5,112.54
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529735,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,NaN,UK,20.65,2011,...,10,9,Q4,Medium,Medium,23,NaN,0.000014,NaN,3918.45
529736,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,NaN,UK,16.52,2011,...,10,9,Q4,Medium,Small,30,NaN,0.000011,NaN,3934.97
529737,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,NaN,UK,4.96,2011,...,10,9,Q4,Low,Small,29,NaN,0.000003,NaN,3939.93
529738,581498,85174,S/4 CACTI CANDLES,1,2011-12-09 10:26:00,10.79,NaN,UK,10.79,2011,...,10,9,Q4,Medium,Small,17,NaN,0.000007,NaN,3950.72


In [ ]:
nonuk = onre[onre['Country']!='UK']

In [ ]:
onre['Group'] = 'uk'
onre['Group'] = 'nonuk'

In [ ]:
uk = uk.reset_index(drop=True)
nonuk = nonuk.reset_index(drop=True)

In [ ]:
com = pd.concat([uk,nonuk],axis=1)
com

,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,...,Hour,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName,Diff,trans,rank,Cumulatives
0,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,UK,77183.60,2011,...,14.0,7.0,Q4,Medium,Medium,31.0,NaN,0.000017,46.0,25.2
1,537215,85124C,GREEN JUICY FRUIT PHOTO FRAME,12,2010-12-05 15:38:00,2.55,12747.0,UK,30.60,2010,...,14.0,7.0,Q4,Medium,Small,33.0,0.0,0.000012,97.5,42.2
2,537215,85124B,BLUE JUICY FRUIT PHOTO FRAME,6,2010-12-05 15:38:00,2.55,12747.0,UK,15.30,2010,...,14.0,7.0,Q4,Medium,Medium,33.0,0.0,0.000027,17.0,81.2
3,537215,84879,ASSORTED COLOUR BIRD ORNAMENT,16,2010-12-05 15:38:00,1.69,12747.0,UK,27.04,2010,...,14.0,7.0,Q4,Medium,Large,23.0,0.0,0.000016,52.0,104.6
4,537215,85062,PEARL CRYSTAL PUMPKIN T-LIGHT HLDR,24,2010-12-05 15:38:00,1.65,12747.0,UK,39.60,2010,...,14.0,7.0,Q4,Medium,Medium,35.0,0.0,0.000010,119.5,119.6
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
485118,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,NaN,UK,20.65,2011,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
485119,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,NaN,UK,16.52,2011,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
485120,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,NaN,UK,4.96,2011,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
485121,581498,85174,S/4 CACTI CANDLES,1,2011-12-09 10:26:00,10.79,NaN,UK,10.79,2011,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


#PART Q: External Data Integration

Task 108: Create a new dataset with customer information (make up data):

CustomerID, CustomerName, MembershipLevel (Bronze/Silver/Gold), JoinDate

In [ ]:
data = {
    'CustomerID': [101, 102, 103, 104, 105],
    'CustomerName': ['Alice Brown', 'Brian Clarke', 'Catherine Lee', 'Daniel Smith', 'Eva Johnson'],
    'MembershipLevel': ['Bronze', 'Silver', 'Gold', 'Silver', 'Gold'],
    'JoinDate': ['2021-05-10', '2022-01-22', '2020-11-15', '2023-03-08', '2021-09-30']
}
customer_df = pd.DataFrame(data)
customer_df

,CustomerID,CustomerName,MembershipLevel,JoinDate
0,101,Alice Brown,Bronze,2021-05-10
1,102,Brian Clarke,Silver,2022-01-22
2,103,Catherine Lee,Gold,2020-11-15
3,104,Daniel Smith,Silver,2023-03-08
4,105,Eva Johnson,Gold,2021-09-30


Task 109: Combine this customer dataset with your sales data to show customer names and membership levels in transactions.

In [ ]:
mem = pd.concat([customer_df,onre],axis=1)
mem

,CustomerID,CustomerName,MembershipLevel,JoinDate,InvoiceNo,StockCode,ProductName,Quantity,InvoiceDate,UnitPrice,...,DayOfWeek,Quarter,RevenueCategory,OrderSize,LenOfProductName,Diff,trans,rank,Cumulatives,Group
0,101.0,Alice Brown,Bronze,2021-05-10,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,...,1,Q4,Medium,Medium,34,NaN,0.000010,162.0,15.30,nonuk
1,102.0,Brian Clarke,Silver,2022-01-22,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,...,1,Q4,Medium,Medium,19,0.0,0.000014,92.5,35.64,nonuk
2,103.0,Catherine Lee,Gold,2020-11-15,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,...,1,Q4,Medium,Medium,30,0.0,0.000015,66.5,57.64,nonuk
3,104.0,Daniel Smith,Silver,2023-03-08,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,...,1,Q4,Medium,Medium,35,0.0,0.000014,92.5,77.98,nonuk
4,105.0,Eva Johnson,Gold,2021-09-30,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,...,1,Q4,Medium,Medium,30,0.0,0.000014,92.5,98.32,nonuk
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529735,NaN,NaN,NaN,NaN,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,...,9,Q4,Medium,Medium,23,NaN,0.000014,NaN,3918.45,nonuk
529736,NaN,NaN,NaN,NaN,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,...,9,Q4,Medium,Small,30,NaN,0.000011,NaN,3934.97,nonuk
529737,NaN,NaN,NaN,NaN,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,...,9,Q4,Low,Small,29,NaN,0.000003,NaN,3939.93,nonuk
529738,NaN,NaN,NaN,NaN,581498,85174,S/4 CACTI CANDLES,1,2011-12-09 10:26:00,10.79,...,9,Q4,Medium,Small,17,NaN,0.000007,NaN,3950.72,nonuk


Task 110: Create a product categories dataset (make up data):

StockCode, Category (e.g., Home, Kitchen, Gift, etc.)

In [ ]:
product_data = {
    'StockCode': ['10001', '10002', '10003', '10004', '10005',
                  '20001', '20002', '20003', '30001', '30002'],
    'Category': [
        'Home', 'Kitchen', 'Gift', 'Toys', 'Home',
        'Stationery', 'Kitchen', 'Gift', 'Electronics', 'Toys'
    ]
}

In [ ]:
product_categories = pd.DataFrame(product_data)
product_categories

,StockCode,Category
0,10001,Home
1,10002,Kitchen
2,10003,Gift
3,10004,Toys
4,10005,Home
5,20001,Stationery
6,20002,Kitchen
7,20003,Gift
8,30001,Electronics
9,30002,Toys


Task 111: Combine the sales data with product categories to analyze which categories generate the most revenue

In [ ]:
comr = pd.concat([onre,product_categories],axis=1)
rev = comr.groupby('Category')['TotalPrice'].sum()
rev

,TotalPrice
Category,
Electronics,11.10
Gift,33.10
Home,35.64
Kitchen,45.84
Stationery,15.30
Toys,74.42


Task 112: Perform an analysis that shows total revenue by MembershipLevel.

In [ ]:
mem.groupby('MembershipLevel')['TotalPrice'].sum()

,TotalPrice
MembershipLevel,
Bronze,15.30
Gold,42.34
Silver,40.68


Task 113: Show which product categories are most popular in each country.

In [ ]:
comr.groupby(['Country','Category'])['TotalPrice'].sum()

Country  Category   
UK       Electronics    11.10
         Gift           33.10
         Home           35.64
         Kitchen        45.84
         Stationery     15.30
         Toys           74.42
Name: TotalPrice, dtype: float64

#PART R: NumPy Operations

Task 114: Extract Quantity and UnitPrice as arrays and perform element-wise multiplication.

In [ ]:
qty = onre['Quantity'].values
unt = onre['UnitPrice'].values

In [ ]:
tot = qty*unt
tot

array([7.71836e+04, 2.52000e+01, 1.70000e+01, ..., 4.96000e+00,
       1.07900e+01, 1.71417e+03])

Task 115: Create an array of 1000 random transaction IDs.

In [ ]:
transID = np.random.randint(1000,size=1000)
transID

array([515, 719, 110, 853, 923, 766, 861, 803, 226, 145, 110, 881, 498,
       442, 250, 689, 666, 782, 798, 155, 561, 538, 679, 953, 329, 894,
       991, 333, 841, 125,  14, 686, 521, 797, 307, 862, 615, 415, 917,
       852, 968, 955, 956, 686, 654, 696, 423, 645, 689, 434, 111, 685,
       103, 164, 334,  89, 580, 851, 295, 172, 842, 916, 304, 562,  29,
       126, 825, 199, 219, 381, 734, 200, 958, 600, 831, 110, 543, 152,
       258, 154, 403, 501, 367, 452, 521, 777,  83, 860, 226,  24, 146,
       602, 306, 674, 283, 435,  19, 791, 546, 885, 637, 272, 296, 376,
       503, 863, 896, 723, 186, 501, 149, 808, 490,  92, 927, 757, 293,
       231, 479, 732, 629, 714, 920, 687,  14, 221, 719, 320,  73, 815,
       976, 690, 254, 843, 779, 464, 118,   8, 258,  76, 430, 984, 128,
       500, 566, 533, 565, 510, 429, 181, 976, 772, 525, 877, 536, 987,
        70, 823, 711,  55, 852, 867, 213,  67,  19, 865, 793,  60, 212,
       549, 863, 717, 860, 792, 296, 973, 349, 761, 385,  28,  6

Task 116: Generate 50 equally spaced price points between £0 and £500.

In [ ]:
price = np.linspace(0,500,50)
price

array([  0.        ,  10.20408163,  20.40816327,  30.6122449 ,
        40.81632653,  51.02040816,  61.2244898 ,  71.42857143,
        81.63265306,  91.83673469, 102.04081633, 112.24489796,
       122.44897959, 132.65306122, 142.85714286, 153.06122449,
       163.26530612, 173.46938776, 183.67346939, 193.87755102,
       204.08163265, 214.28571429, 224.48979592, 234.69387755,
       244.89795918, 255.10204082, 265.30612245, 275.51020408,
       285.71428571, 295.91836735, 306.12244898, 316.32653061,
       326.53061224, 336.73469388, 346.93877551, 357.14285714,
       367.34693878, 377.55102041, 387.75510204, 397.95918367,
       408.16326531, 418.36734694, 428.57142857, 438.7755102 ,
       448.97959184, 459.18367347, 469.3877551 , 479.59183673,
       489.79591837, 500.        ])

Task 117: Create a matrix representing 12 months and 7 product categories filled with zeros.

In [ ]:
monmat = np.zeros((12,7))
monmat

array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

Task 118: Extract quantities greater than 100 from the Quantity array.

In [ ]:
onre[onre['Quantity']>100]['Quantity']

,Quantity
60250,74215
144671,240
33306,120
33308,120
33310,144
...,...
517000,132
518246,184
525596,105
525615,205


Task 119: Calculate the square root of all UnitPrice values.

In [ ]:
sq_up = np.sqrt(onre['UnitPrice'])
sq_up

,UnitPrice
60250,1.019804
14637,1.449138
14638,2.061553
14639,1.802776
14640,0.806226
...,...
529735,2.032240
529736,2.032240
529737,2.227106
529738,3.284814


Task 120: Find the 25th, 50th, and 75th percentiles of TotalPrice.

In [ ]:
p25 = np.percentile(onre['TotalPrice'], 25)
p50 = np.percentile(onre['TotalPrice'], 50)
p75 = np.percentile(onre['TotalPrice'], 75)

In [ ]:
p25

np.float64(3.75)

In [ ]:
p50

np.float64(9.9)

In [ ]:
p75

np.float64(17.700000000000003)

Task 121: Round all UnitPrice values to the nearest integer.

In [ ]:
onre['UnitPrice_Rounded'] = onre['UnitPrice'].round()

Task 122: Find positions where Quantity is maximum.

In [ ]:
max_positions = onre.index[onre['Quantity'] == onre['Quantity'].max()].tolist()
max_positions

[528623]

Task 123: Calculate the correlation between Quantity and UnitPrice.

In [ ]:
correlation = onre[['Quantity', 'UnitPrice']].corr().loc['Quantity', 'UnitPrice']
correlation

np.float64(-0.003772542661636776)

Task 124: Normalize the TotalPrice column (subtract mean, divide by standard deviation).

In [ ]:
onre['TotalPrice_Normalized'] = (
    (onre['TotalPrice'] - onre['TotalPrice'].mean()) /
     onre['TotalPrice'].std()
)

Task 125: Split the Quantity array into 4 equal parts.

In [ ]:
qty = onre['Quantity'].values

parts = np.array_split(qty, 4)
parts

[array([74215,    12,     4, ...,     2,     4,     4]),
 array([100,   8,  16, ...,   4,   2,   1]),
 array([1, 2, 1, ..., 1, 1, 1]),
 array([1, 1, 1, ..., 1, 1, 1])]

#PART S: Export and Documentation

Task 126: Save your cleaned dataset to a CSV file

In [ ]:
onre.to_csv("cleaned_sales_data.csv", index=False)

Task 127: Save your cleaned dataset to an Excel file.

In [ ]:
onre.to_excel("cleaned_sales_data.xlsx", index=False)

Task 128: Create an Excel file with multiple sheets:

Sheet 1: Full cleaned data
Sheet 2: Country-wise summary
Sheet 3: Monthly revenue trends
Sheet 4: Top 50 customers

In [ ]:
onre['Sales'] = onre['Quantity'] * onre['UnitPrice']

In [ ]:
sheet1 = onre

In [ ]:
sheet2 = onre.groupby('Country').agg({
    'Sales': 'sum',
    'Quantity': 'sum'
}).reset_index()

In [ ]:
onre['InvoiceDate'] = pd.to_datetime(onre['InvoiceDate'])
sheet3 = onre.groupby(onre['InvoiceDate'].dt.to_period('M'))['Sales'].sum().reset_index()
sheet3['InvoiceDate'] = sheet3['InvoiceDate'].astype(str)

In [ ]:
sheet4 = (
    onre.groupby('CustomerID')['Sales']
    .sum()
    .reset_index()
    .sort_values(by='Sales', ascending=False)
    .head(50)
)

In [ ]:
!pip install xlsxwriter
# Create Excel with multiple sheets
with pd.ExcelWriter("analysis_report.xlsx", engine='xlsxwriter') as writer:

    # Sheet 1
    sheet1.to_excel(writer, sheet_name="Full Data", index=False)

    # Sheet 2
    sheet2.to_excel(writer, sheet_name="Country Summary", index=False)

    # Sheet 3
    sheet3.to_excel(writer, sheet_name="Monthly Revenue", index=False)

    # Sheet 4
    sheet4.to_excel(writer, sheet_name="Top 50 Customers", index=False)

print("Excel file created successfully!")

Excel file created successfully!


Task 129: Export the pivot table from Task 100 to Excel.

In [ ]:
import numpy as np
import pandas as pd
onre['InvoiceDate'] = pd.to_datetime(onre['InvoiceDate'])

pivot_100 = onre.pivot_table(
    index='Country',
    columns=onre['InvoiceDate'].dt.month,
    values='Sales',
    aggfunc='sum'
)

pivot_100.to_excel("pivot_100.xlsx", index=True)


Task 130: Create a text file documenting all the insights you discovered from this analysis.

In [ ]:
text = """
# Analysis Insights Summary

## 1. Sales Insights
- Identified highest sales months and seasonal patterns.
- Determined top-selling products based on revenue.
- Found that certain Outlet Establishment Years outperform others in total sales.

## 2. Customer Insights
- Identified top 50 customers based on total revenue generated.
- Customers with more frequent purchases tend to have higher lifetime value.
- Calculated average days between purchases for each customer.

## 3. Product Insights
- Determined top 20 products by revenue.
- Found products sold in more than 10 different countries.
- Identified most expensive and cheapest products.

## 4. Country-Level Insights
- Calculated total revenue by country.
- Countries with highest average order values identified.
- Monthly sales trends visualized for each country.

## 5. Time-Series Trends
- Monthly revenue trends show clear growth patterns.
- Specific months show peak demand depending on regions.

## 6. Data Cleaning Insights
- Missing CustomerID rows removed.
- Negative or zero Quantity values were cleaned.
- Missing descriptions filled with 'Unknown Product'.

## 7. NumPy Calculations
- Performed element-wise operations using Quantity and UnitPrice.
- Computed square roots, splits, and array transformations for deeper analysis.

## 8. Pivot Table Insights
- Monthly sales by country provided comparative performance views.
- Pivot exported for further reporting.

This document summarizes the key findings and analysis results obtained during the entire project.
"""

# --- Save to a text file using standard file handling ---
with open("analysis_insights.txt", "w", encoding="utf-8") as file:
    file.write(text)

print("File saved successfully!")


File saved successfully!


In [ ]:
file